mithril-common/src/database/version_checker.rs

```diff
@@ -0,0 +1,294 @@
+            });
+
+        assert_eq!(semver, version.semver.to_string());
+    }
+
+    fn create_sqlite_file(name: &str) -> PathBuf {
+        let filepath = std::env::temp_dir().join(name);
+
+        if filepath.exists() {
+            std::fs::remove_file(filepath.as_path()).unwrap();
+        }
+
+        filepath
+    }
+
+    fn get_table_whatever_column_count(filepath: &PathBuf) -> i64 {
+        let connection = Connection::open(filepath).unwrap();
+        let sql = "select count(*) as column_count from pragma_table_info('whatever');";
+        let column_count = connection
+            .prepare(sql)
+            .unwrap()
+            .into_cursor()
+            .bind(&[])
+            .unwrap()
+            .next()
+            .unwrap()
+            .unwrap()
+            .get::<i64, _>(0);
+
+        column_count
+    }
+
+    #[test]
+    fn test_database_version_checker() {
+        let filepath = create_sqlite_file("test_1.sqlite3");
+        let mut db_checker = DatabaseVersionChecker::new(
+            slog_scope::logger(),
+            ApplicationNodeType::Aggregator,
+            filepath.clone(),
+        );
+        db_checker.add_migration(SqlMigration {
+            version: Version::from_str("1.0.10").unwrap(),
+            alterations: "select false".to_string(),
+        });
+        db_checker.check("1.0.0").unwrap();
+        check_database_version(&filepath, "0.0.0");
+
+        db_checker.check("1.0.0").unwrap();
+        check_database_version(&filepath, "0.0.0");
+
+        db_checker.check("1.1.0").unwrap();
+        check_database_version(&filepath, "1.0.10");
+
+        db_checker.check("1.0.12").unwrap();
+        check_database_version(&filepath, "1.0.10");
+
+        db_checker.check("1.0.9").unwrap_err();
+    }
+
+    #[test]
+    fn test_upgrade_with_migration() {
+        let filepath = create_sqlite_file("test_2.sqlite3");
+        let mut db_checker = DatabaseVersionChecker::new(
+            slog_scope::logger(),
+            ApplicationNodeType::Aggregator,
+            filepath.clone(),
+        );
+
+        // The order of the migrations is intentionnaly made in reverse to
+        // ensure the version semver is the ordering criteria when applied.
+        let alterations = "alter table whatever add column thing_content text; update whatever set thing_content = 'some content'";
+        let migration = SqlMigration {
+            version: Version::from_str("1.1.0").unwrap(),
+            alterations: alterations.to_string(),
+        };
+        db_checker.add_migration(migration);
+
+        let alterations = "create table whatever (thing_id integer); insert into whatever (thing_id) values (1), (2), (3), (4);";
+        let migration = SqlMigration {
+            version: Version::from_str("1.0.5").unwrap(),
+            alterations: alterations.to_string(),
+        };
+        db_checker.add_migration(migration);
+
+        db_checker.check("1.0.0").unwrap();
```

Comment:
check → apply
save the number of migrations played an automatically apply new migrations
remove `app_version` and create an new table `db_version` with db_version as int instead of text.